File: heaps.py
from math import (floor)


class MaxBinaryHeap:
    data = 'Added Value'

    def __init__(self):
        self.values = [41, 39, 33, 18, 27, 12]

    def insert(self, value):
        self.values.append(value)
        self.siftUp()

    def siftUp(self):
        index = len(self.values) - 1
        ELEMENT = self.values[index]
        while index > 0:
            parentIndex = floor((index - 1) / 2)
            parentElement = self.values[parentIndex]
            if ELEMENT <= parentElement:
                break
            if ELEMENT > parentElement:
                self.values[parentIndex] = ELEMENT
                self.values[index] = parentElement
                index = parentIndex

    def siftDown(self):
        root = self.values[0]
        end = self.values.pop()
        self.values[0] = end
        index = 0
        length = len(self.values)
        element = self.values[0]
        while True:
            leftChildIndex = 2 * index + 1
            rightChildIndex = 2 * index + 2
            leftChild = None
            rightChild = None
            swap = None
            if leftChildIndex < length:
                leftChild = self.values[leftChildIndex]
                if leftChild > element:
                    swap = leftChildIndex
            if rightChildIndex < length:
                rightChild = self.values[rightChildIndex]
                if (swap is None and rightChild > element) or (swap is not None and rightChild > leftChild):
                    swap = rightChildIndex
            if swap is None:
                break
            self.values[index] = self.values[swap]
            self.values[swap] = element
            index = swap
        return root


maxB = MaxBinaryHeap()
maxB.insert(55)
print(maxB.values)
maxB.siftDown()
print(maxB.values)


def log(name):
    if 'data' in name:
        print(name)
    else:
        print('Not Exist')


log('Mohamed')

File: heaps.js
class MaxBinaryHeap {
  constructor() {
    this.values = [41, 39, 33, 18, 27, 12];
  }

  insert(value) {
    this.values.push(value);
    this.bubbleUp();
  }

  bubbleUp() {
    let index = this.values.length - 1;
    const ELEMENT = this.values[index];
    while (index > 0) {
      let parentIndex = Math.floor((index - 1) / 2);
      let parentElement = this.values[parentIndex];
      if (ELEMENT <= parentElement) break;
      if (ELEMENT > parentElement) {
        this.values[parentIndex] = ELEMENT;
        this.values[index] = parentElement;
        index = parentIndex;
      }
    }
  }

  siftDown() {
    var root = this.values[0];
    var end = this.values.pop();
    this.values[0] = end;
    var element = this.values[0];
    var index = 0;
    var leftChildIndex = 2 * index + 1;
    var rightChildIndex = 2 * index + 2;
    var leftChild;
    var rightChild;
    var swap = null;
    while (true) {
      if (leftChildIndex) {
        leftChild = this.values[leftChildIndex];
        if (leftChild > element) {
          swap = leftChildIndex;
        }
      }
      if (rightChildIndex) {
        rightChild = this.values[rightChildIndex];
        if (
          (swap === null && rightChild > element) ||
          (swap !== null && rightChild > leftChild)
        ) {
          swap = rightChildIndex;
        }
      }
      if (swap == null) {
        break;
      }
      this.values[index] = this.values[swap];
      this.values[swap] = element;
      index = swap;
    }
    return root;
  }
}

const maxBH = new MaxBinaryHeap();
maxBH.insert(55);
console.log(maxBH.values);
